src/plugins/data/public/query/query_string/dataset_service/lib/index_pattern_type.ts

```diff
@@ -105,12 +105,23 @@ const fetchIndexPatterns = async (client: SavedObjectsClientContract): Promise<D
     perPage: 100,
   });
 
-  // Get all unique data source ids
+  // Get all unique data source ids from both references and index pattern IDs
   const datasourceIds = Array.from(
     new Set(
       resp.savedObjects
-        .filter((savedObject) => savedObject.references.length > 0)
-        .map((savedObject) => savedObject.references.find((ref) => ref.type === 'data-source')?.id)
+        .map((savedObject) => {
+          // First try to get from references
+          const refDataSourceId = savedObject.references.find((ref) => ref.type === 'data-source')
+            ?.id;
+          if (refDataSourceId) {
+            return refDataSourceId;
+          }
+          // If not in references, check if the ID contains :: (namespaced format)
+          if (savedObject.id.includes('::')) {
```

Comment:
Not related to this PR but weird why don't we keep the reference way for data source for dataset? Detecting special chars to parse data source id is introducing a dependency on how data source based index pattern id gets generated.